documentation/project-docs/breaking-change-guidelines.md

```diff
@@ -0,0 +1,81 @@
+# SDK Breaking Change and Diagnostic Guidelines
+
+This document provides guidelines for introducing new diagnostics or breaking changes to the .NET SDK, which configuration knobs are available, and what criteria should guide the decision around severity, timeline for introduction, and deprecation of diagnostics.
+
+## Overall procedure
+
+In general, we want to make updating the .NET SDK as smooth as possible for developers. This means:
+
+* introducing new changes in a staged/gradual way
+* trying to tie opinionated analyzers/diagnostics to a mechanism that requires explicit user opt-in
+* providing a way to opt out of a change entirely
+
+## Kinds of changes
+
+There are many kinds of breaking changes that can ship in the .NET SDK, such as:
+
+* New MSBuild warnings and errors (props/targets)
+* New NuGet warnings and errors. 
+  * For example, NuGet Audit.
+* Roslyn Analyzers and CodeFixes
+  * This includes trimming/ILLink analyzers and codefixes
+* Behavioral/implementation changes
+  * MSBuild engine changes like MSBuild Server
+  * Implementation changes for MSBuild Tasks
+  * NuGet Restore algorithm enhancements
+  * Changes to DotNet CLI grammar
+  * Changes to defaults in CLI flags that impact behavior
+    * For example, `--configuration` flag defaulting to `Release` instead of `Debug`
+
+## Configuration Knobs
+
+The following knobs are available to enable/disable these changes (some may not apply to all kinds of changes):
+
+* [TargetFramework](https://learn.microsoft.com/en-us/dotnet/standard/frameworks)
+* [SdkAnalysisLevel](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#sdkanalysislevel)
+* EditorConfig (for Analyzers)
+* AnalysisLevel (for Analyzers/CodeFixes)
+* WarningLevel
+
+## Deciding how to Surface Changes
+
+### Implement changes in an informational/non-blocking way initially
+
+What this means will vary change-to-change. For example, for a change expressed as an Analyzer or MSBuild diagnostic, consider
+Informational level severities initially. For a behavioral change on a CLI, consider an informational message written to the
```

Comment:
theoretically, this is also breaking but we have to draw a line somewhere and I think if some script is reading and depending on the stderr output, that seems like a reasonable line to draw.

Reply:
@rainersigwald - Based on your experience, should changes int the output be considered breaking?

Reply:
Changes in output can certainly be breaking. MSBuild has had to demote some high-importance info messages (e.g. https://github.com/dotnet/msbuild/pull/9228) because including location and/or a code made them look "too much" like warnings for enough users.

Reply:
Addressed.